src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1020 @@
+
+	D_DEBUG(DB_TRACE, "Querying parity for stripe: %lu, offset: %lu\n",
+		entry->ae_cur_stripe.as_stripenum,
+		iter_param.ip_recx.rx_idx);
+
+	/* Query the parity */
+	rc = vos_iterate(&iter_param, VOS_ITER_RECX, false, &anchors,
+			 agg_recx_iter_pre_cb, NULL, entry, NULL);
+	/* entry->ae_par_extent.ape_epoch has been set to the parity extent's
+	 * epoch
+	 */
+	if (rc != 0)
+		goto out;
+
+	D_DEBUG(DB_TRACE, "Par query: epoch: %lu, offset: %lu, length: %lu\n",
+		entry->ae_par_extent.ape_epoch,
+		entry->ae_par_extent.ape_recx.rx_idx,
+		entry->ae_par_extent.ape_recx.rx_nr);
+
+	if (entry->ae_par_extent.ape_epoch > entry->ae_cur_stripe.as_hi_epoch &&
+	    entry->ae_par_extent.ape_epoch != ~(0ULL)) {
+		/* Parity newer than data and holes, so delete the replicas.
+		 */
+		update_vos = true;
+		write_parity = false;
+		goto out;
+	}
+
+	if ((entry->ae_par_extent.ape_epoch == ~(0ULL) &&
+	     agg_stripe_is_filled(entry, false)) ||
+	     agg_stripe_is_filled(entry, true)) {
+		/* Replicas constitute a full stripe. */
+		rc = agg_encode_local_parity(entry);
+		goto out;
+	}
+
+	if (entry->ae_par_extent.ape_epoch == ~(0ULL)) {
+		/* No parity, partial-stripe worth of replica, nothing to do */
+		update_vos = false;
+		goto out;
+	}
+out:
+	if (update_vos && rc == 0) {
+		entry->ae_cur_stripe.as_suffix_ext = agg_get_carry_under(entry);
+		rc = agg_update_vos(entry, write_parity);
+	}
+
+	agg_clear_extents(entry);
+	return rc;
+}
+
+/* returns the subrange of the RECX iterator's returned recx that lies within
+ * the current stripe.
+ */
+static daos_off_t
+agg_in_stripe(struct ec_agg_entry *entry, daos_recx_t *recx)
+{
+	unsigned int		len = entry->ae_oca->u.ec.e_len;
+	unsigned int		k = entry->ae_oca->u.ec.e_k;
+	daos_off_t		stripe = recx->rx_idx / (len * k);
+	daos_off_t		stripe_end = (stripe + 1) * len * k;
+
+	if (recx->rx_idx + recx->rx_nr > stripe_end)
+		return stripe_end - recx->rx_idx;
+	else
+		return recx->rx_nr;
+}
+
+/* Iterator call back sub-function for handling data extents.
+ */
+static int
+agg_data_extent(vos_iter_entry_t *entry, struct ec_agg_entry *agg_entry,
+		unsigned int *acts)
+{
+	struct ec_agg_extent	*extent = NULL;
+	daos_off_t		 cur_stripenum, this_stripenum;
+	int			 rc = 0;
+
+	this_stripenum = agg_stripenum(agg_entry, entry->ie_recx.rx_idx);
+	if (this_stripenum != agg_entry->ae_cur_stripe.as_stripenum) {
+		/* Iterator has reached next stripe */
+		if (agg_entry->ae_cur_stripe.as_extent_cnt) {
+			cur_stripenum = agg_entry->ae_cur_stripe.as_stripenum;
+			rc = agg_process_stripe(agg_entry);
+			if (rc)
+				D_ERROR("Process stripe returned "DF_RC"\n",
+					DP_RC(rc));
+		/* Error leaves data covered by replicas vulnerable to vos
+		 * delete, so don't advance coordination epoch.
+		 */
+			rc = 0;
+			if (cur_stripenum <
+			    agg_entry->ae_cur_stripe.as_stripenum &&
+			agg_entry->ae_cur_stripe.as_stripenum <
+			this_stripenum) {
+				/* Handle holdover stripe */
```

Comment:
It's bit hard to follow here. I suppose it's for processing the extents spanning stripes? Why it can't be handled by above agg_process_stripe() call?

Reply:
agg_process_stripe is invoked once we the iterator encounters a subsequent stripe. If there an extent that spanned the stripe, the suffix of it carrying over to the next stripe was save as extent in the extent list.

If there are more extents in the next stripe, we don't know about them in the previous call to process stripe.

There are two cases for the spanning extent.  1) next stripe has more extents. 2) the spanning extent is the only extent in the next stripe.

Reply:
Maybe I can have a better view if you could explain how the fully covered extents & overlapping extents (whose logical visible recx != physical recx) cases are processed?

Reply:
The iterator finds all visible extents in the data range: rx_idx: 0 -- (1 << 63) -1. As it encounter extents the code in agg_data_extent() adds the extents to a linked list. When the iterator submits an extent with start offset beyond the "current stripe" (the stripe represented by the extent list), it invokes agg_process_stripe(), if there is a now prior stripe to be processed (i.e, the list in not empty).

agg_process_stripe() does the necessary work for the stripe and then clears the extents from the list.  The exception is if the last extent spanned to the next stripe. In this case the spanning extent is retained in the list, but its start offset is moved to be the start of the next (now current) stripe. The length of the "prefix" part, the range in the prior stripe, is retained so the spanning extent can be removed from VOS correctly.

After agg_process_stripe() returns, the code in agg_data_extent checks to see if the spanning extent was the only extent in the following stripe. In this case, there is an additional call to agg_process_stripe() to handle this subsequent stripe.


Assume stripe k is had two extents: they fill the stripe and second extent spans to the next stripe, partially filling it. There is no data in the stripe after this stripe, and the next data is several stripe further in the address range.

Stripe k:

[|-------------------------------||---------------------------------------------]

Stripe k + 1:

[-------------------------|                                                                            ]

Strike k + m (m > 1):

[|----------------------------------------------------------------------------|]

The iterator encounters the first two extents that start in stripe k and adds them to the list.  Then the third extent (starting in stripe k + m) is passed to agg_data_extent().  Since this stripenum is > stripenum of the extent list, stripe k is processed. Following this, the cur_stripenum for the stripe has been advanced, and the remaining portion of the spanning extent has been retained as the only extent in stripe k + 1. Since the stripenum of the extent passed to agg_data_extent() is > k +1, a second call to agg_process_stripe is invoked on the stripe k + 1.

After stripe k+1 is processsed, and the extent list is cleared, the extent for stripe k + m is added to the list and the processing of extents proceeds.

Fully covered extents are not retrieved by the iterator, only visible extents.  Holes are handled in the follow-up, full-featured patch.

Reply:
I see, thanks for the explanation. So if I'm understanding correct, the second call to agg_process_stripe() could be missed in agg_akey_post(), right? If that's a defect, it could be fixed along with the extent removal fix.